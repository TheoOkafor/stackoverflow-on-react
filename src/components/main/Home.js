import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from "react-router-dom";
import PropTypes from 'prop-types';

import { fetchQuestions } from '../../actions/questionActions';
import Card from './Card';


class Home extends Component {
  componentDidMount() {
    this.props.fetchQuestions();
  }

  shouldComponentUpdate(nextProps){
    const check = this.props.newQuestion === nextProps.newQuestion;
    if(!check) {
      this.props.questions.unshift(nextProps.newQuestion);
      return true;
    }
    return true;
  }

  render() {
    return (
    <div>
      { 
        this.props.questions.map( question => {
          let answer = '';
          let time = '';
          let username = '';
          let answersUserid = '';
          let currAnswers = question.answers? question.answers : [];
          let numAnswers = question.numAnswers || 0;

          // To display accepted answer or any random answer.
          let acceptedAnswer = currAnswers.filter( answer => {
            return answer.accepted;
          });
          if (acceptedAnswer.length > 0){
            answer = acceptedAnswer[0].body;
            time =  new Date(acceptedAnswer[0].timesubmitted).toDateString();
            username = acceptedAnswer[0].username;
            answersUserid = acceptedAnswer[0].userid;
          } else if (numAnswers > 0) {
            let randomIndex = Math.floor(Math.random() * currAnswers.length);
            answer = currAnswers[randomIndex].body;
            time = new Date(currAnswers[randomIndex].timesubmitted)
              .toDateString();
            username = currAnswers[randomIndex].username;
            answersUserid = currAnswers[randomIndex].userid;
          } else {
              answer = 'No answers yet';
              time = '';
              username = '';
          }
          return(
          <Card  key={question.id || question.questionid}>
            <div>
              <h3 className="qs-title">
                <Link to= {`/questions/${question.id || question.questionid}`} 
                className="question">{question.title}</Link>
              </h3>

              <div>
                <h5 className="person-answer"><Link className="inherit"
                to= {`/users/${answersUserid}`}>
                {username}</Link><br />
                  <small>Answered: 	<span>{time}</span></small>
                </h5>
              </div>
              
              <p className="answer">
                {answer}
              </p>
              <h6>Answers: <span>{numAnswers}</span></h6>
            </div>
          </Card>
          );
        })
      }
    </div>
    );
  }
}

const mapStateToProps = state => ({
  questions: state.questions.questions,
  newQuestion: state.questions.question,
})

Home.propTypes = {
  fetchQuestions: PropTypes.func.isRequired,
  questions: PropTypes.array,
  newQuestion: PropTypes.object,
};

export default connect(mapStateToProps, { fetchQuestions })(Home);
