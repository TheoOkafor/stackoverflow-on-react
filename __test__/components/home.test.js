import React from 'react';
import renderer from 'react-test-renderer';
import Home from '../../src/components/main/Home';
// import data from '../utilities/data';

test.skip('Test that the Home displays', () => {
  const component = renderer.create(
    <Home />
  );
  expect(component).toMatchSnapshot();
});
