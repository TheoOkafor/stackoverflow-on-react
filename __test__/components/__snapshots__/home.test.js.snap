// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcceptAnswer component Test that the Home displays 1`] = `<div />`;

exports[`AcceptAnswer component Test that the Home displays 2`] = `
<div>
  <div
    className="card undefined"
  >
    <div>
      <h3
        className="qs-title"
      >
        <a
          className="question"
          href="/questions/1"
          onClick={[Function]}
        >
          why do yo shake?
        </a>
      </h3>
      <div>
        <h5
          className="person-answer"
        >
          <a
            className="inherit"
            href="/users/4"
            onClick={[Function]}
          >
            teddy01
          </a>
          <br />
          <small>
            Answered: 
            <span>
              Wed Nov 21 2018
            </span>
          </small>
        </h5>
      </div>
      <p
        className="answer"
      >
        worddy
      </p>
      <h6>
        Answers: 
        <span>
          7
        </span>
      </h6>
    </div>
  </div>
  <div
    className="card undefined"
  >
    <div>
      <h3
        className="qs-title"
      >
        <a
          className="question"
          href="/questions/2"
          onClick={[Function]}
        >
          why do yo shake?
        </a>
      </h3>
      <div>
        <h5
          className="person-answer"
        >
          <a
            className="inherit"
            href="/users/4"
            onClick={[Function]}
          >
            teddy01
          </a>
          <br />
          <small>
            Answered: 
            <span>
              Wed Nov 21 2018
            </span>
          </small>
        </h5>
      </div>
      <p
        className="answer"
      >
        worddy
      </p>
      <h6>
        Answers: 
        <span>
          7
        </span>
      </h6>
    </div>
  </div>
  <div
    className="card undefined"
  >
    <div>
      <h3
        className="qs-title"
      >
        <a
          className="question"
          href="/questions/3"
          onClick={[Function]}
        >
          why do yo shake?
        </a>
      </h3>
      <div>
        <h5
          className="person-answer"
        >
          <a
            className="inherit"
            href="/users/"
            onClick={[Function]}
          >
            
          </a>
          <br />
          <small>
            Answered: 
            <span>
              
            </span>
          </small>
        </h5>
      </div>
      <p
        className="answer"
      >
        No answers yet
      </p>
      <h6>
        Answers: 
        <span>
          0
        </span>
      </h6>
    </div>
  </div>
  <div
    className="card undefined"
  >
    <div>
      <h3
        className="qs-title"
      >
        <a
          className="question"
          href="/questions/4"
          onClick={[Function]}
        >
          why do yo shake?
        </a>
      </h3>
      <div>
        <h5
          className="person-answer"
        >
          <a
            className="inherit"
            href="/users/"
            onClick={[Function]}
          >
            
          </a>
          <br />
          <small>
            Answered: 
            <span>
              
            </span>
          </small>
        </h5>
      </div>
      <p
        className="answer"
      >
        No answers yet
      </p>
      <h6>
        Answers: 
        <span>
          0
        </span>
      </h6>
    </div>
  </div>
  <div
    className="card undefined"
  >
    <div>
      <h3
        className="qs-title"
      >
        <a
          className="question"
          href="/questions/5"
          onClick={[Function]}
        >
          why do yo shake?
        </a>
      </h3>
      <div>
        <h5
          className="person-answer"
        >
          <a
            className="inherit"
            href="/users/"
            onClick={[Function]}
          >
            
          </a>
          <br />
          <small>
            Answered: 
            <span>
              
            </span>
          </small>
        </h5>
      </div>
      <p
        className="answer"
      >
        No answers yet
      </p>
      <h6>
        Answers: 
        <span>
          0
        </span>
      </h6>
    </div>
  </div>
  <div
    className="card undefined"
  >
    <div>
      <h3
        className="qs-title"
      >
        <a
          className="question"
          href="/questions/6"
          onClick={[Function]}
        >
          why do yo shake?
        </a>
      </h3>
      <div>
        <h5
          className="person-answer"
        >
          <a
            className="inherit"
            href="/users/"
            onClick={[Function]}
          >
            
          </a>
          <br />
          <small>
            Answered: 
            <span>
              
            </span>
          </small>
        </h5>
      </div>
      <p
        className="answer"
      >
        No answers yet
      </p>
      <h6>
        Answers: 
        <span>
          0
        </span>
      </h6>
    </div>
  </div>
</div>
`;
